pkg/trace/writer/trace.go

```diff
@@ -127,183 +120,202 @@ func NewTraceWriter(
-}
-
-// Run starts the TraceWriter.
-func (w *TraceWriter) Run() {
-	if w.syncMode {
-		w.runSync()
-	} else {
-		w.runAsync()
-	}
-}
-
-func (w *TraceWriter) runAsync() {
-	t := time.NewTicker(w.tick)
-	defer t.Stop()
-	defer close(w.Serialize)
-	defer close(w.stop)
+func (w *TraceWriter) reporter() {
+	tck := time.NewTicker(w.tick)
+	defer w.wg.Done()
 	for {
 		select {
-		case pkg := <-w.In:
-			w.addSpans(pkg)
+		case <-tck.C:
+			w.report()
 		case <-w.stop:
-			w.drainAndFlush()
 			return
-		case <-t.C:
-			w.report()
-			w.flush()
 		}
 	}
 }
 
-func (w *TraceWriter) runSync() {
-	defer close(w.Serialize)
-	defer close(w.stop)
-	defer close(w.flushChan)
+func (w *TraceWriter) timeFlush() {
+	defer w.wg.Done()
 	for {
 		select {
-		case pkg := <-w.In:
-			w.addSpans(pkg)
-		case notify := <-w.flushChan:
-			w.drainAndFlush()
-			notify <- struct{}{}
+		case <-w.flushTicker.C:
+			func() {
+				w.mu.Lock()
+				defer w.mu.Unlock()
+				w.flush()
+			}()
 		case <-w.stop:
-			w.drainAndFlush()
 			return
 		}
 	}
 }
 
+// Stop stops the TraceWriter and attempts to flush whatever is left in the senders buffers.
+func (w *TraceWriter) Stop() {
+	log.Debug("Exiting trace writer. Trying to flush whatever is left...")
+	close(w.stop)
+	// Wait for encoding/compression to complete on each payload,
+	// and submission to senders
+	w.wg.Wait()
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.flush()
+	stopSenders(w.senders)
+}
+
 // FlushSync blocks and sends pending payloads when syncMode is true
 func (w *TraceWriter) FlushSync() error {
 	if !w.syncMode {
 		return errors.New("not flushing; sync mode not enabled")
 	}
 	defer w.report()
 
-	notify := make(chan struct{}, 1)
-	w.flushChan <- notify
-	<-notify
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.flush()
 	return nil
 }
 
-func (w *TraceWriter) addSpans(pkg *SampledChunks) {
-	w.stats.Spans.Add(pkg.SpanCount)
-	w.stats.Traces.Add(int64(len(pkg.TracerPayload.Chunks)))
-	w.stats.Events.Add(pkg.EventCount)
-
+// appendChunks adds sampled chunks to the current payload, and in the case the payload
+// is full, returns a finished payload which needs to be written out.
+func (w *TraceWriter) appendChunks(pkg *SampledChunks) []*pb.TracerPayload {
+	var toflush []*pb.TracerPayload
+	w.mu.Lock()
+	defer w.mu.Unlock()
 	size := pkg.Size
 	if size+w.bufferedSize > MaxPayloadSize {
 		// reached maximum allowed buffered size
-		w.flush()
+		// reset the buffer so we can add our payload and defer a flush.
+		toflush = w.tracerPayloads
+		w.resetBuffer()
 	}
 	if len(pkg.TracerPayload.Chunks) > 0 {
 		log.Tracef("Writer: handling new tracer payload with %d spans: %v", pkg.SpanCount, pkg.TracerPayload)
 		w.tracerPayloads = append(w.tracerPayloads, pkg.TracerPayload)
 	}
 	w.bufferedSize += size
+	return toflush
 }
 
-func (w *TraceWriter) drainAndFlush() {
-outer:
-	for {
-		select {
-		case pkg := <-w.In:
-			w.addSpans(pkg)
-		default:
-			break outer
-		}
+// WriteChunks writes and serializes the provided chunks
```

Comment:
With the current synchronous structure, how does _writing_ differ from _serializing_? Can a single verb represent the whole operation, or do we want to keep the explicit separation between them?

Reply:
Yeah in this case "writing" is synonymous with serializing I believe. This is admittedly confusing since there's still a separate "sending" step which happens asynchronously (except for serverless). 
Is this mostly thinking we should update the comment to drop the word serializes? Maybe something like
`// WriteChunks serializes the provided chunks, enqueueing them to be sent`?

Reply:
Yes that works and seems clearer to me 👍